Name Registry contract (#238)
FuelLabs/sway-applications: name-registry/SPECIFICATION.md

Table of Contents
- [Overview](#overview)
- [Use Cases](#use-cases)
  - [Actions that users are able to perform](#actions-that-users-are-able-to-perform)
  - [Information that is presented to users](#information-that-is-presented-to-users)

# Overview

This document provides an overview of the application.

It outlines the use cases, i.e. desirable functionality, in addition to requirements for the smart contract and the user interface.

# Use Cases

This section contains general information about the functionality of the application and thus does not touch upon any technical aspects.

If you are interested in a functional overview then this is the section for you.

## Actions that users are able to perform

This sub-section details what a user is able to do e.g. click a button and "x, y, z" happens.

### `register()`

A user can register a name for themselves

> 1. If the name is not already registered

Reviewer: This is slightly misleading. A name can be "registered" but expired. Does that mean that it is still registered? I would reword this

2. If the payment is sufficient for the duration
3. If the payment is in the correct asset

### `extend()`

> Any user can extend the registration duration of a given name

Reviewer: Should mention that the name must be registered because otherwise it will revert

1. If the payment is sufficient for the duration
2. If the payment is in the correct asset

### `set_identity()`

> The owner of the name can change the identity to which the name resolves

Reviewer: ```suggestion
Allows the owner to change the resolving identity
```

Unsure about this suggestions but I think it's better than the current one

### `set_owner()`

> The owner of the name can change the owner of the name

Reviewer: ```suggestion
Allows the owner to transfer ownership of the registered name
```

> The owner of the name should be able to change the owner of the name

Author: I used the fundraiser as a template, the specification in the fundraiser uses "should" everywhere so I thought thats what we were going for?
